Services/Models/InternalDeviceState.cs

```diff
@@ -0,0 +1,247 @@
+                }
+                else
+                {
+                    this.properties.Add(key, value);
+                    log.Debug("Added new device property", () => new { key, value });
+                }
+
+                // mark the device for reported properties updates to the IoT Hub
+                this.PropertyChanged = true;
+            }
+        }
+
+        /// <summary>
+        /// Retrieve all simulation state values as a read only dictionary
+        /// </summary>
+        public Dictionary<string, object> GetState()
+        {
+            // lock properties as mulitple scripts may try to access key at the same time.
+            lock (this.simulationState)
+            {
+                // TODO investigate returning a read only dictionary to
+                // enforce updates through the set property
+                return new Dictionary<string, object>(this.simulationState);
+            }
+        }
+
+        /// <summary>
+        /// Update the current simulation state
+        /// </summary>
+        public void SetState(Dictionary<string, object> newState)
+        {
+            // lock properties as mulitple scripts may try to access key at the same time.
+            lock (this.simulationState)
+            {
+                this.simulationState = newState;
+            }
+        }
+
+        /// <summary>
+        /// Returns true if key exists
+        /// </summary>
+        public bool HasStateValue(string key)
+        {
+            // lock properties as mulitple scripts may try to access key at the same time.
+            lock (this.simulationState)
+            {
+                return this.simulationState.ContainsKey(key);
+            }
+        }
+
+        /// <summary>
+        /// Retrieve simulated device state value.
+        /// </summary>
+        public object GetStateValue(string key)
+        {
+            // lock properties as mulitple scripts may try to access key at the same time.
+            lock (this.simulationState)
+            {
+                return this.simulationState[key];
+            }
+        }
+
+        /// <summary>
+        /// Set the simulation state value with the given key, adds new value if key does not exist.
+        /// </summary>
+        public void SetStateValue(string key, object value)
+        {
+            // lock properties as mulitple scripts may try to access key at the same time.
+            lock (this.simulationState)
+            {
+                if (this.simulationState.ContainsKey(key))
+                {
+                    this.simulationState[key] = value;
+                }
+                else
+                {
+                    this.simulationState.Add(key, value);
+                }
+            }
+        }
+
+        private Dictionary<string, object> SetupTelemetry(DeviceModel deviceModel)
+        {
+            // put telemetry properties in state
```

Comment:
note, the state does not contain "only" telemetry data, also simulation state, e.g. see the "simulation_state" field here https://github.com/Azure/device-simulation-dotnet/blob/azure-iot-pcs-simulation/Services/data/devicemodels/chiller-01.json